### PR TITLE
add balance check contract to repo

### PR DESCRIPTION
**Pull request by poiresel**
## Description

Adding the batch balance and approval checker to the airswap-protocols repo
Goal will be to deploy this contract to the test chains and work with the deltaBalances logic within AirSwap.js

Quick description:

- [X] New features

## Tests
```
  27 passing (13s)

---------------------|----------|----------|----------|----------|----------------|
File                 |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
---------------------|----------|----------|----------|----------|----------------|
 contracts/          |      100 |       95 |      100 |      100 |                |
  BalanceChecker.sol |      100 |       95 |      100 |      100 |                |
  Imports.sol        |      100 |      100 |      100 |      100 |                |
---------------------|----------|----------|----------|----------|----------------|
All files            |      100 |       95 |      100 |      100 |                |
---------------------|----------|----------|----------|----------|----------------|
```

### CODE REVIEW

File: source/balances/contracts/BalanceChecker.sol [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+    * @param tokenAddresses address[]
+    * @return uint256[] token allowances array if possible
+    */
+  function allAllowancesForManyAccounts(
+      address[] calldata userAddresses,
+      address spenderAddress,
+      address[] calldata tokenAddresses
+    ) external view returns (uint256[] memory) {
+    uint256[] memory allowances = new uint256[](
+      tokenAddresses.length * userAddresses.length);
```
Comment (alicevhenshaw) [inline source/balances/contracts/BalanceChecker.sol]: indentation on this line and the 2 above is weird

File: source/balances/contracts/BalanceChecker.sol [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    */
+  function walletAllowances(
+    address userAddress,
+    address spenderAddress,
+    address[] calldata tokenAddresses
+  ) external view returns (uint256[] memory) {
+    require(tokenAddresses.length > 0);
+    uint256[] memory allowances = new uint256[](tokenAddresses.length);
+
+    for(uint256 i = 0; i< tokenAddresses.length; i++) {
```
Comment (alicevhenshaw) [inline source/balances/contracts/BalanceChecker.sol]: space before `<`

File: source/balances/contracts/BalanceChecker.sol [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    */
+  function walletBalances(
+    address userAddress,
+    address[] calldata tokenAddresses
+  ) external view returns (uint256[] memory) {
+    require(tokenAddresses.length > 0);
+    uint256[] memory balances = new uint256[](tokenAddresses.length);
+
+    for(uint256 i = 0; i < tokenAddresses.length; i++){
+      if( tokenAddresses[i] != address(0x0) ) {
```
Comment (alicevhenshaw) [inline source/balances/contracts/BalanceChecker.sol]: weird spacing on the brackets. For `if` `while` and `for` we formally do:
`if (...) {`

File: source/balances/contracts/BalanceChecker.sol [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    require(success, "ETH_WITHDRAW_FAILED");
+  }
+
+  /**
+    * @notice Allow owner to withdraw stuck tokens from contract
+    * @param tokenAddress address
+    * @param amount uint256
+    */
+  function withdrawToken(
+  address tokenAddress,
```
Comment (alicevhenshaw) [inline source/balances/contracts/BalanceChecker.sol]: indentation

File: source/balances/contracts/BalanceChecker.sol [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    return balances;
+  }
+
+  /**
+    * @notice Checks WETH balances
+    * @param wethAddress address WETH addresses in case of several
+    * @param userAddresses address[]
+    * @return uint256[]
+    */
+  function allWETHbalances(
```
Comment (alicevhenshaw) [inline source/balances/contracts/BalanceChecker.sol]: whats the need for a WETH-specific function? Can this function not be generic for all IERC20s?

File: source/balances/README.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
+# Delegate
```
Comment (dmosites) [inline source/balances/README.md]: Balances

File: source/balances/README.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+## Resources
+
+- Docs → https://docs.airswap.io/
+- Website → https://www.airswap.io/
+- Blog → https://blog.airswap.io/
+- Support → https://support.airswap.io/
+
+## Deploys
+
+[BalanceChecker](./contracts/BalanceChecker.sol) is deployed. See [deploys.json](./deploys.json) for latest contract deployments on Goerli (`5`) and Kovan (`42`).
```
Comment (dmosites) [inline source/balances/README.md]: Please add mainnet (`1`) and rinkeby (`4`)

**Pull request merged**